src/bin/pgcopydb/ld_apply.c

```diff
@@ -1819,3 +1857,183 @@ stream_apply_read_lsn_tracking(StreamApplyContext *context)
+ * metadata filename.
+ */
+static bool
+parseTxnMetadataFile(const char *filename, LogicalMessageMetadata *metadata)
+{
+	/* store xid as it will be overwritten while parsing metadata */
+	uint32_t xid = metadata->xid;
+
+	if (xid == 0)
+	{
+		log_error("BUG: parseTxnMetadataFile is called with "
+				  "transaction xid: %lld", (long long) xid);
+		return false;
+	}
+
+	char *txnMetadataContent = NULL;
+	long size = 0L;
+
+	if (!read_file(filename, &txnMetadataContent, &size))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	JSON_Value *json = json_parse_string(txnMetadataContent);
+
+	if (!parseMessageMetadata(metadata, txnMetadataContent, json, true))
+	{
+		/* errors have already been logged */
+		json_value_free(json);
+		return false;
+	}
+
+	json_value_free(json);
+	free(txnMetadataContent);
+
+	if (metadata->txnCommitLSN == InvalidXLogRecPtr ||
+		metadata->xid != xid ||
+		IS_EMPTY_STRING_BUFFER(metadata->timestamp))
+	{
+		log_error("Failed to parse metadata for transaction metadata file "
+				  "%s: %s", filename, txnMetadataContent);
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ *  computeTxnMetadataFilename computes the file path for transaction metadata
+ *  based on its transaction id
+ */
+static bool
+computeTxnMetadataFilename(uint32_t xid, const char *dir, char *filename)
+{
+	if (dir == NULL)
+	{
+		log_error("BUG: computeTxnMetadataFilename is called with "
+				  "directory: NULL");
+		return false;
+	}
+
+	if (xid == 0)
+	{
+		log_error("BUG: computeTxnMetadataFilename is called with "
+				  "transaction xid: %lld", (long long) xid);
+		return false;
+	}
+
+	sformat(filename, MAXPGPATH, "%s/%lld.json", dir, (long long) xid);
+
+	return true;
+}
+
+
+/*
+ * writeTxnCommitMetadata writes the transaction metadata to a file in the given
+ * directory
+ */
+static bool
+writeTxnCommitMetadata(LogicalMessageMetadata *mesg, const char *dir)
+{
+	char txnfilename[MAXPGPATH] = { 0 };
+
+	if (mesg->action != STREAM_ACTION_COMMIT)
+	{
+		log_error("BUG: writeTxnCommitMetadata is called with "
+				  "action: %c", mesg->action);
+		return false;
+	}
+
+	if (!computeTxnMetadataFilename(mesg->xid, dir, txnfilename))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	log_debug("stream_write_commit_metadata_file: writing transaction "
+			  "metadata file \"%s\" with commit lsn %X/%X",
+			  txnfilename,
+			  LSN_FORMAT_ARGS(mesg->lsn));
+
+	char contents[BUFSIZE] = { 0 };
+
+	sformat(contents, BUFSIZE,
+			"{\"xid\":%lld,\"commit_lsn\":\"%X/%X\",\"timestamp\":\"%s\"}\n",
+			(long long) mesg->xid,
+			LSN_FORMAT_ARGS(mesg->lsn),
+			mesg->timestamp);
+
+	/* write the metadata to txnfilename */
+	if (!write_file(contents, strlen(contents), txnfilename))
+	{
+		log_error("Failed to write file \"%s\"", txnfilename);
+		return false;
+	}
+
+	log_debug("Wrote transaction metadata file for "
+			  "txn %lld: %s",
+			  (long long) mesg->xid,
+			  txnfilename);
```

Comment:
Filename should be in double-quotes in logs, always, allowing copy/paste even with strange filenames containing spaces or other strange stuff.